core/amber/src/main/python/core/models/test_RTableExecutor.py

```diff
@@ -0,0 +1,494 @@
+        tuple_3 = Tuple({"col1": None, "col2": None, "col3": None})
+        return [tuple_1, tuple_2, tuple_3]
+
+    @pytest.fixture
+    def pandas_target_df_simple(self):
+        data = {
+            "Name": ["Alice", "Bob", "Charlie"],
+            "Age": [25, 30, 35],
+            "City": ["New York", "Los Angeles", "Chicago"],
+        }
+        df = pandas.DataFrame(data)
+        return df
+
+    @pytest.fixture
+    def target_tuples_simple(self, pandas_target_df_simple):
+        tuples = []
+        for index, row in pandas_target_df_simple.iterrows():
+            tuples.append(Tuple(row))
+        return tuples
+
+    @pytest.fixture
+    def source_executor_simple(self):
+        return """
+        function() {
+            df <- data.frame(
+              Name = c("Alice", "Bob", "Charlie"),
+              Age = c(25, 30, 35),
+              City = c("New York", "Los Angeles", "Chicago")
+              )
+            return (df)
+        }
+        """
+
+    @pytest.fixture
+    def udf_executor_simple(self):
+        return """
+        function(input, port) {
+            return (input)
+        }
+        """
+
+    @pytest.fixture
+    def udf_executor_simple_extract_row(self):
+        return """
+        function(input, port) {
+            bob_row <- input[input$Name == "Bob", ]
+            return (bob_row)
+        }
+        """
+
+    @pytest.fixture
+    def udf_executor_simple_update_row(self):
+        return """
+        function(input, port) {
+            input[input$Name == "Bob", "Age"] <- 18
+            return (input)
+        }
+        """
+
+    @pytest.fixture
+    def udf_executor_simple_add_row(self):
+        return """
+        function(input, port) {
+            new_row <- list(Name = "Test", Age = 0, City = "Irvine")
+            new_df <- rbind(input, new_row)
+            return (new_df)
+        }
+        """
+
+    @pytest.fixture
+    def source_executor_df_fail(self):
+        # This Source UDF should raise a TypeError since it cannot
+        # be converted into a Table-like object
+        return """
+        function() {
+            glm_model <- glm(mpg ~ wt, data = mtcars, family = gaussian)
+            return (glm_model)
+        }
+        """
+
+    @pytest.fixture
+    def target_tuples_like_type(self):
+        tuple_1 = Tuple({"C.1": 1, "C.2": 2, "C.3": 3})
+        tuple_2 = Tuple({"C.1": 11, "C.2": 12, "C.3": 13})
+        return [tuple_1, tuple_2]
+
+    @pytest.fixture
+    def source_executor_df_like_type(self):
+        return """
+        function() {
+            mdat <- matrix(c(1,2,3, 11,12,13), nrow = 2, ncol = 3, byrow = TRUE,
+                dimnames = list(c("row1", "row2"),
+                c("C.1", "C.2", "C.3")))
+            return (mdat)
+        }
+        """
+
+    @pytest.fixture
+    def udf_executor_df_like_type(self):
+        return """
+        function(input, port) {
+            return (input)
+        }
+        """
```

Comment:
seems to be a duplicate?

Reply:
Removed this and used the one from udf_executor_simple_return. I wanted to see if the UDF could handle echoing data types that were similar to the R dataframe.